src/discovery/discoveryLogic.ts

```diff
@@ -1,90 +1,247 @@
+export async function loadCommunityTypeIndexes (store:LiveStore, user:NamedNode): Promise<TypeIndexScope[][]> {
   const preferencesFile = await loadPreferences(store, user)
-  if (preferencesFile) {
+  if (preferencesFile) { // For now, pick up communities as simple links from the preferences file.
     const communities = store.each(user, ns.solid('community'), undefined, preferencesFile as NamedNode)
-    const communityTypeIndexesPromise = communities.map(async community => await loadTypeIndexesFor(store, community as NamedNode))
-    const result1 = Promise.all(communityTypeIndexesPromise)
-    // const result2 = Promise.all(result1)
-    // const flat = result2.flat()
-    return result1
-    // const communityTypeIndexes = await Promise.all(communityTypeIndexesPromise)
-      /*
-    let result = [] as TypeIndex[]
-    for(const community of communities) {
-      result = result.concat(await loadTypeIndexesFor(store, community as NamedNode)) as TypeIndex[] // @@ how oto make functional with async?
+    // console.log('loadCommunityTypeIndexes communities: ',communities)
+    let result = []
+    for (const org of communities) {
+      result = result.concat(await loadTypeIndexesFor(store, org as NamedNode))
     }
-    */
-    // return communityTypeIndexesPromise.resolve()
+    // const communityTypeIndexesPromises = communities.map(async community => await loadTypeIndexesFor(store, community as NamedNode))
+    // const result1 = Promise.all(communityTypeIndexesPromises)
+    return result
   }
-  return []
+  return [] // No communities
 }
 
 export async function loadAllTypeIndexes (store:LiveStore, user:NamedNode) {
   return (await loadTypeIndexesFor(store, user)).concat((await loadCommunityTypeIndexes(store, user)).flat())
 }
 
-/*
-export async function getAppInstances (store:LiveStore, klass: NamedNode) {
+// Utility: remove duplicates from Array of NamedNodes
 
+export function uniqueNodes (arr: NamedNode[]): NamedNode[] {
+  const uris = arr.map(x => x.uri)
+  const set = new Set(uris)
+  const uris2 = Array.from(set)
+  const arr2 = uris2.map(u => new NamedNode(u))
+  return arr2 // Array.from(new Set(arr.map(x => x.uri))).map(u => sym(u))
 }
-*/
+
+export async function getScopedAppsFrommIndex (store, scope, theClass: NamedNode) {
+  // console.log(`getScopedAppsFrommIndex agent ${scope.agent} index: ${scope.index}` )
+  const index = scope.index
+  const registrations = store.each(undefined, ns.solid('forClass'), theClass, index)
+  // console.log('    registrations', registrations )
+
+  // In practice it looks as though existing code does not put in the type explicitly so can't do this filter. Discuss.
+  // const relevant = registrations.filter(reg => store.holds(reg, ns.rdf('type'), ns.solid('TypeRegistration'), index))
+  // console.log('    relevant', relevant )
+  const relevant = registrations
+
+  const directInstances = registrations.map(reg => store.any(reg as NamedNode, ns.solid('instance'), null, index))
```

Comment:
So we expect there to be only one instance or container per type. Correct?

Reply:
Yup